src/cephadm/cephadm.py

```diff
@@ -3231,15 +3238,22 @@ def fetch_custom_config_files(ctx: CephadmContext) -> List[Dict[str, Any]]:
     return []
 
 
-def fetch_tcp_ports(ctx: CephadmContext) -> List[int]:
+def fetch_tcp_ports(ctx: CephadmContext) -> Tuple[List[int], Dict[str, str]]:
```

Comment:
Instead of returning a tuple, how about returning a new type (something like `PortSelection` perhaps)? The new type would have methods (or properties) for port listing or port-to-ip mappings. This type can encapsulate the knowledge/logic around the daemons port configuration.

As an additional benefit, instead of adding a new parameter to many functions you can just change the type of the existing `daemon_ports` argument. Then only functions that *use* daemon_ports need to change, other ones that just pass the argument along can continue to be the same code (modulo the type signature update).

Reply:
OK, I was wondering if the info in port_ips was a superset we could just convert to port_ips everywhere. Maybe that'd be a good idea in the future but I won't push for it in this PR. My other comment still stands, I think encapsulating these items together (since they are still logically coupled) would be a good thing.

Reply:
I think there's definitely an argument to be made that we want to only have 1 object we pass down and tcp_ports and port_ips should be merged somehow into one thing. That would require a more in depth change as we'd have to then modify the handling of tcp_ports in the binary and how tcp_ports was generated to begin with in the mgr module. I can at least take a look at what that sort of change would look like and at the very least handle the suggestion from your first comment.

Reply:
I've modified this to use the EndPoint class that was already defined in cephadm and combines a port and ip. I also ended up needing an additional commit to fix some error handling in port_in_use/attempt_bind. Not sure how that didn't come up the first time I was trying this, probably a messy test env after so many failed haproxy deployments and upgrades.